### PR TITLE
[Feature] Support RTMDet

### PR DESCRIPTION
**Pull request by hhaAndroid**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Please describe the motivation of this PR and the goal you want to achieve through this PR.

## Modification

Please briefly describe what modification is made in this PR.

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repos?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects, like MMDetection or MMClassification.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: mmyolo/models/dense_heads/rtmdet_head.py [hunk 1]
```diff
@@ -0,0 +1,683 @@
...
+            else:
+                reg_dist = self.rtm_reg[idx](reg_feat) * stride
+            cls_scores.append(cls_score)
+            bbox_preds.append(reg_dist)
+        return tuple(cls_scores), tuple(bbox_preds)
+
+
+@MODELS.register_module()
+class RTMDetHead(YOLOv5Head):
+    """YOLOXHead head used in `YOLOX <https://arxiv.org/abs/2107.08430>`_.
```
Comment (PeterH0323) [inline mmyolo/models/dense_heads/rtmdet_head.py:201]: Should be change to RTMDet paper ?

File: mmyolo/models/necks/cspnext_pafpn.py [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+        norm_cfg: ConfigType = dict(type='BN', momentum=0.03, eps=0.001),
+        act_cfg: ConfigType = dict(type='SiLu', inplace=True),
+        init_cfg: OptMultiConfig = dict(
+            type='Kaiming',
+            layer='Conv2d',
+            a=math.sqrt(5),
+            distribution='uniform',
+            mode='fan_in',
+            nonlinearity='leaky_relu')
+    ) -> None:
```
Comment (PeterH0323) [inline mmyolo/models/necks/cspnext_pafpn.py:60]:
```suggestion
    ):
```

File: mmyolo/models/dense_heads/rtmdet_head.py [hunk 3]
```diff
@@ -0,0 +1,683 @@
...
+
+    def get_targets(self,
+                    cls_scores: Tensor,
+                    bbox_preds: Tensor,
+                    anchor_list: List[List[Tensor]],
+                    valid_flag_list: List[List[Tensor]],
+                    batch_gt_instances: InstanceList,
+                    batch_img_metas: List[dict],
+                    batch_gt_instances_ignore: OptInstanceList = None,
+                    unmap_outputs=True):
```
Comment (PeterH0323) [inline mmyolo/models/dense_heads/rtmdet_head.py:456]:
```suggestion
                    unmap_outputs=True) -> Union[None, list]:
```

File: mmyolo/models/dense_heads/rtmdet_head.py [hunk 4]
```diff
@@ -0,0 +1,683 @@
...
+            pos_bbox_weight = bbox_targets.new_tensor(0.)
+
+        return loss_cls, loss_bbox, assign_metrics.sum(), pos_bbox_weight.sum()
+
+    def loss_by_feat(self,
+                     cls_scores: List[Tensor],
+                     bbox_preds: List[Tensor],
+                     batch_gt_instances: InstanceList,
+                     batch_img_metas: List[dict],
+                     batch_gt_instances_ignore: OptInstanceList = None):
```
Comment (PeterH0323) [inline mmyolo/models/dense_heads/rtmdet_head.py:375]:
```suggestion
                     batch_gt_instances_ignore: OptInstanceList = None) -> dict:
```

File: mmyolo/models/dense_heads/rtmdet_head.py [hunk 5]
```diff
@@ -0,0 +1,683 @@
...
+            None,
+            batch_img_metas=batch_img_metas,
+            cfg=cfg,
+            rescale=rescale,
+            with_nms=with_nms)
+
+    def loss_by_feat_single(self, cls_score: Tensor, bbox_pred: Tensor,
+                            labels: Tensor, label_weights: Tensor,
+                            bbox_targets: Tensor, assign_metrics: Tensor,
+                            stride: List[int]):
```
Comment (PeterH0323) [inline mmyolo/models/dense_heads/rtmdet_head.py:310]: Should add return type hint.

File: mmyolo/models/dense_heads/rtmdet_head.py [hunk 6]
```diff
@@ -0,0 +1,683 @@
...
+
+    def _get_targets_single(self,
+                            cls_scores: Tensor,
+                            bbox_preds: Tensor,
+                            flat_anchors: Tensor,
+                            valid_flags: Tensor,
+                            gt_instances: InstanceData,
+                            img_meta: dict,
+                            gt_instances_ignore: Optional[InstanceData] = None,
+                            unmap_outputs=True):
```
Comment (PeterH0323) [inline mmyolo/models/dense_heads/rtmdet_head.py:549]:
```suggestion
                            unmap_outputs=True) -> list:
```

**Review by PeterH0323 — CHANGES_REQUESTED** [review]

File: mmyolo/models/necks/cspnext_pafpn.py [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+    def build_reduce_layer(self, idx: int) -> nn.Module:
+        """build reduce layer.
+
+        Args:
+            idx (int): layer idx.
+
+        Returns:
+            nn.Module: The reduce layer.
+        """
+        if idx == 2:
```
Comment (RangeKing) [inline mmyolo/models/necks/cspnext_pafpn.py:89]:
```suggestion
        if idx in (1, 2):
```
According to https://github.com/open-mmlab/mmdetection/blob/a3cf5368fd3725c9848479f54675eb70bb0fa00b/mmdet/models/necks/cspnext_pafpn.py#L70-L78

File: mmyolo/models/necks/cspnext_pafpn.py [hunk 8]
```diff
@@ -0,0 +1,196 @@
...
+                    expand_ratio=self.expand_ratio,
+                    conv_cfg=self.conv_cfg,
+                    norm_cfg=self.norm_cfg,
+                    act_cfg=self.act_cfg),
+                self.conv(
+                    self.in_channels[idx - 1],
+                    self.in_channels[idx - 2],
+                    kernel_size=1,
+                    norm_cfg=self.norm_cfg,
+                    act_cfg=self.act_cfg))
```
Comment (RangeKing) [inline mmyolo/models/necks/cspnext_pafpn.py:142]:
```suggestion
        return CSPLayer(
            self.in_channels[idx - 1] * 2,
            self.in_channels[idx - 1],
            num_blocks=self.num_csp_blocks,
            add_identity=False,
            use_cspnext_block=True,
            expand_ratio=self.expand_ratio,
            conv_cfg=self.conv_cfg,
            norm_cfg=self.norm_cfg,
            act_cfg=self.act_cfg)
```

File: configs/rtmdet/cspnext_imagenet_pretrain/README.md [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+
+```shell
+pip install mmcls>=1.0.0rc0
+```
+
+## Prepare Dataset
+
+To pre-train on ImageNet, you need to prepare the dataset first. Please refer to the [guide](https://mmclassification.readthedocs.io/en/1.x/user_guides/dataset_prepare.html#imagenet).
+
+## How to Train
```
Comment (Nioolek) [inline configs/rtmdet/cspnext_imagenet_pretrain/README.md:25]: Whether the `mim` command can be added

Comment (hhaAndroid) [reply configs/rtmdet/cspnext_imagenet_pretrain/README.md:25]: No need to use mim, it's easier

**Pull request merged**